File: lib/features/show-types.js
/*
 * decaffeinate suggestions:
 * DS102: Remove unnecessary code created because of implicit returns
 * DS207: Consider shorter variations of null checks
 * Full docs: https://github.com/decaffeinate/decaffeinate/blob/master/docs/suggestions.md
 */
const {bufferPositionFromMouseEvent, pixelPositionFromMouseEvent} = require('../utils');
const SubAtom = require('sub-atom');
const {goToPosition} = require('./go-to');
const log = require('loglevel').getLogger('ensime.show-types');

// This one lives as one per file for all instances with an instanceLookup.
class ShowTypes {
  constructor(editor, clientLookup) {
    this.editor = editor;
    this.clientLookup = clientLookup;
    this.disposables = new SubAtom;
    this.locked = false;

    this.editorView = atom.views.getView(this.editor);
    this.editorElement = this.editorView;

    atom.config.observe('Ensime.enableTypeTooltip', enabled => {
      if(enabled) {
        this.disposables.add(this.editorElement, 'mousemove', '.scroll-view', e => {
          this.clearExprTypeTimeout();
          return this.exprTypeTimeout = setTimeout((() => {
            return this.showExpressionType(e);
          }
          ), 100);
        });

        this.disposables.add(this.editorElement, 'mouseout', '.scroll-view', e => {
          return this.clearExprTypeTimeout();
        });

        this.disposables.add(this.editor.onDidDestroy(() => {
          return this.deactivate();
        })
        );
          
        return this.disposables.add(atom.config.observe('Ensime.richTypeTooltip', richTypeTooltip => {
          this.richTypeTooltip = richTypeTooltip;
          
      }));
      } else {
        return this.deactivate();
      }
    });
  }
        // @disposables.dispose()

  // get expression type under mouse cursor and show it
  showExpressionType(e) {
    const {formatTypeAsString, formatTypeAsHtml, typeConstructorFromName} = require('../atom-formatting');
    
    if ((this.marker != null) || this.locked) { return; }

    const pixelPt = pixelPositionFromMouseEvent(this.editor, e);
    const bufferPt = bufferPositionFromMouseEvent(this.editor, e);
    
    const offset = this.editor.getBuffer().characterIndexForPosition(bufferPt);

    const client = this.clientLookup();
    return (client != null ? client.getSymbolAtPoint(this.editor.getPath(), offset).then(msg => {
      if (this.marker != null) {
        this.marker.destroy();
      }
      
      if(msg.type.fullName === "<none>") { return; }
    
      this.marker = this.editor.markBufferPosition(bufferPt);
      if(this.marker) {
        const typeFormatter =
          this.richTypeTooltip ? formatTypeAsHtml : formatTypeAsString;
        
        const TypeHoverElement = require('../views/type-hover-element');
        const element = new TypeHoverElement().initialize(typeFormatter(msg.type));
        
        if (this.domListener != null) {
          this.domListener.destroy();
        }
        const DOMListener = require('dom-listener');
        this.domListener = new DOMListener(element);
        this.domListener.add("a", 'click', event => {
          const a = event.target;
          const qualifiedName = decodeURIComponent(a.dataset.qualifiedName);
          log.debug("asking for symbol by name: ", qualifiedName);
          return client.symbolByName(qualifiedName).then(response => {
            if(response.declPos) {
              goToPosition(response.declPos);
              return this.unstickAndHide();
            }
          });
        });
          
          
        this.overlayDecoration = this.editor.decorateMarker(this.marker, {
          type: 'overlay',
          item: element,
          class: "ensime"
        });
        
        if (this.stickCommand != null) {
          this.stickCommand.dispose();
        }
        return this.stickCommand = atom.commands.add('atom-workspace', "ensime:lock-type-hover", () => {
          this.locked = true;
          this.stickCommand.dispose();
          if (this.unstickCommand != null) {
            this.unstickCommand.dispose();
          }
          return this.unstickCommand = atom.commands.add('atom-workspace', "core:cancel", () => {
            return this.unstickAndHide();
        });
      });
      }
    }).catch(function(err) {}) : undefined);
  }
      // Do nothing, this happens when hovering on "stuff"
      
  unstickAndHide() {
    this.unstickCommand.dispose();
    this.locked = false;
    return this.hideExpressionType();
  }

  deactivate() {
    this.clearExprTypeTimeout();
    this.disposables.dispose();
    if (this.stickCommand != null) {
      this.stickCommand.dispose();
    }
    if (this.unstickCommand != null) {
      this.unstickCommand.dispose();
    }
    return (this.domListener != null ? this.domListener.destroy() : undefined);
  }

  // helper function to hide tooltip and stop timeout
  clearExprTypeTimeout() {
    if (this.exprTypeTimeout != null) {
      clearTimeout(this.exprTypeTimeout);
      this.exprTypeTimeout = null;
    }
    return this.hideExpressionType();
  }

  hideExpressionType() {
    if (this.locked || !this.marker) { return; }
    if (this.marker != null) {
      this.marker.destroy();
    }
    this.marker = null;
    return (this.openerDisposable != null ? this.openerDisposable.dispose() : undefined);
  }
}

module.exports = ShowTypes;

File: src/main/scala/features/show-types.scala
package org.ensime.atom.features

import scala.scalajs.js
import scala.scalajs.js.annotation._

import org.ensime.atom.Utils
import org.ensime.atom.AtomFormatting
import org.ensime.atom.external.{
  AtomEnvironment,
  DisplayMarker,
  Disposable,
  DOMListener,
  Ensime,
  LogLevel,
  SubAtom,
  TextEditor,
  Timeout,
  Timers
}
import org.ensime.atom.views.TypeHoverElement

@JSExportAll
@JSExportTopLevel("ShowTypes")
class ShowTypes(editor: TextEditor,  clientLookup: js.Function0[js.UndefOr[Ensime]]) {
  val log = LogLevel.getLogger("ensime.show-types")
  val disposables = new SubAtom()
  var locked = false
  val editorElement = AtomEnvironment.views.getView(editor)

  var marker: js.UndefOr[DisplayMarker] = js.undefined

  var exprTypeTimeout: js.UndefOr[Timeout] = js.undefined
  var openerDisposable: js.UndefOr[js.Dynamic] = js.undefined

  var richTypeTooltip: js.UndefOr[js.Dynamic] = js.undefined

  var unstickCommand: js.UndefOr[Disposable] = js.undefined
  var stickCommand: js.UndefOr[Disposable] = js.undefined

  var domListener: js.UndefOr[DOMListener] = js.undefined

  AtomEnvironment.config.observe[Boolean]("Ensime.enableTypeTooltip", enabled => {
    if(enabled){
      disposables.add(editorElement, "mousemove", ".scroll-view", (e) => {
        clearExprTypeTimeout()
        exprTypeTimeout = Timers.setTimeout(() => showExpressionType(e), 100)
      })

      disposables.add(editorElement, "mouseout", ".scroll-view", (e) => clearExprTypeTimeout())

      disposables.add(editor.onDidDestroy(() => deactivate()))

      disposables.add(AtomEnvironment.config.observe[js.Dynamic]("Ensime.richTypeTooltip", richTypeTooltip_ =>
        richTypeTooltip = richTypeTooltip_
      ))
    }else{
      deactivate()
    }
  })



  def showExpressionType(e: js.Any): Unit = {
    if(locked || marker.isDefined)
      return ()
    else {
      val pixelPt = Utils.pixelPositionFromMouseEvent(editor, e)
      val bufferPt = Utils.bufferPositionFromMouseEvent(editor, e)

      val offset = editor.getBuffer().characterIndexForPosition(bufferPt)

      clientLookup().foreach{client =>
        client.getSymbolAtPoint(editor.getPath(), offset).andThen{msg =>
          marker.foreach(_.destroy())

          if(msg.symbolType.fullName.asInstanceOf[String] == "<none>"){
            // return
          } else {
            marker = editor.markBufferPosition(bufferPt)
            val typeFormatter =
              if(richTypeTooltip.isDefined) AtomFormatting.formatTypeAsHtml(_) else AtomFormatting.formatTypeAsString(_)

            val element = new TypeHoverElement().initialize(typeFormatter(msg.symbolType))
            // domListener.foreach(_.destroy())
            // domListener = new DOMListener(element)
            // domListener.get.add("a", "click", { event =>
            //   val a = event.target
            //   val qualifiedName = URIUtils.decodeURIComponent(a.dataset.qualifiedName)
            //   log.debug("asking for symbol by name: ", qualifiedName)
            //   client.symbolByName(qualifiedName).andThen{ response =>
            //     if(response.declPos)
            //       Goto.goToPosition(response.declPos)
            //       unstickAndHide()
            //   }
            // })


            val overlayDecoration = editor.decorateMarker(marker.get, js.Dynamic.literal(
              `type` = "overlay",
              item = element,
              `class` = "ensime"
            ))

            stickCommand.foreach(_.dispose())
            stickCommand = AtomEnvironment.commands.add("atom-workspace", "ensime:lock-type-hover", {_ =>
              locked = true
              stickCommand.foreach(_.dispose())
              unstickCommand.foreach(_.dispose())
              unstickCommand = AtomEnvironment.commands.add("atom-workspace", "core:cancel", _ => unstickAndHide())
            })
          }
        }.recover{(err) =>
          // Do nothing, this happens when hovering on "stuff"
        }
      }
    }
  }

  def unstickAndHide(): Unit = {
    unstickCommand.foreach(_.dispose())
    locked = false
    hideExpressionType()
  }

  def deactivate(): Unit = {
    clearExprTypeTimeout()
    disposables.dispose()
    stickCommand.foreach(_.dispose())
    unstickCommand.foreach(_.dispose())
    domListener.foreach(_.destroy())
  }

  // helper function to hide tooltip and stop timeout
  def clearExprTypeTimeout(): Unit = {
    exprTypeTimeout.foreach(timeout =>
      Timers.clearTimeout(timeout)
    )
    exprTypeTimeout = js.undefined
    hideExpressionType()
  }

  def hideExpressionType(): Unit = {
    if(locked || marker.isEmpty)
      return ()
    else
    marker.foreach(_.destroy())
    marker = js.undefined
    openerDisposable.foreach(_.dispose())
  }

}
